writer.go

```diff
@@ -19,28 +17,45 @@ type Writer struct {
 	framer    Framer
 	formatter Formatter
 
+	mu   sync.RWMutex // guards conn
 	conn serverConn
 }
 
+// getConn provides access to the internal conn, protected by a mutex. The
+// conn is threadsafe, so it can be used while unlocked, but we want to avoid
+// race conditions on grabbing a reference to it.
+func (w *Writer) getConn() serverConn {
+	w.mu.RLock()
+	defer w.mu.RUnlock()
+	return w.conn
+}
+
+// setConn updates the internal conn, protected by a mutex.
+func (w *Writer) setConn(c serverConn) {
+	w.mu.Lock()
+	w.conn = c
+	w.mu.Unlock()
+}
+
 // connect makes a connection to the syslog server.
-// It must be called with w.mu held.
-func (w *Writer) connect() (err error) {
-	if w.conn != nil {
+func (w *Writer) connect() (serverConn, error) {
+	conn := w.getConn()
+	if conn != nil {
 		// ignore err from close, it makes sense to continue anyway
-		w.conn.close()
-		w.conn = nil
+		conn.close()
+		w.setConn(nil)
 	}
 
-	var conn serverConn
 	var hostname string
+	var err error
 	dialer := w.getDialer()
 	conn, hostname, err = dialer.Call()
 	if err == nil {
-		w.conn = conn
+		w.setConn(conn)
 		w.hostname = hostname
 	}
 
-	return
+	return conn, err
```

Comment:
That makes sense. My only misgiving is that the pattern of `(good, nil), (nil, err)` is used elsewhere.  In my mind, it does seem clearer to explicitly set `nil, err` in the presence of the error or return `con, nil` when all has gone well. This is simply because the condition of receiving an initialized yet defective connection shouldn't be possible.

But (!), I'll leave the decision of how to move forward up to you. My reasoning is solely based on consistency whereas yours is rooted in pragmatism. I don't think either one is strictly better than the other. :)